xclim/indices/_agro.py

```diff
@@ -884,14 +885,63 @@ def water_budget(
     return out
 
 
+def _preprocess_spx(da, freq, window):
+    _, base, _, _ = parse_offset(freq or xarray.infer_freq(da.time))
+    try:
+        group = {"D": "time.dayofyear", "M": "time.month"}[base]
+    except KeyError():
+        raise ValueError(f"Standardized index with frequency `{freq}` not supported.")
+    if freq:
+        da = da.resample(time=freq).mean(keep_attrs=True)
+    if window > 1:
+        da = da.rolling(time=window).mean(skipna=False, keep_attrs=True)
+    return da, group
+
+
+def _compute_spx_fit_params(da, cal_range, freq, window, dist, method, group=None):
+    # "WPM" method doesn't seem to work for gamma or pearson3
+    dist_and_methods = {"gamma": ["ML", "APP"], "fisk": ["ML", "APP"]}
+    if dist not in dist_and_methods:
+        raise NotImplementedError(f"The distribution `{dist}` is not supported.")
+    if method not in dist_and_methods[dist]:
+        raise NotImplementedError(
+            f"The method `{method}` is not supported for distribution `{dist}`."
+        )
+    if group is None:
+        da, group = _preprocess_spx(da, freq, window)
+
+    if cal_range:
+        da = da.sel(time=slice(cal_range[0], cal_range[1]))
+
+    if uses_dask(da) and len(da.chunks[da.get_axis_num("time")]) > 1:
+        warnings.warn(
+            "The input data is chunked on time dimension and must be fully rechunked to"
+            " run `fit` on groups ."
+            " Beware, this operation can significantly increase the number of tasks dask"
+            " has to handle.",
+            stacklevel=2,
+        )
+        da = da.chunk({"time": -1})
+    params = da.groupby(group).map(fit, (dist, method))
+    params.attrs.update(
+        {
+            "Calibration period": str(cal_range),
+            "Distribution": dist,
+            "Method": method,
+        }
+    )
+    return params
+
+
```

Comment:
I'm still unsure about this structure. I wanted to get the fitting function out of `spi`. Before, `spi` had 2 output types, either I asked for `params`, or for actual `spi` values. Mixing output types (both unitless, but still conceptually different with different shapes for the output DataArray) is not ideal.

However, the functions that then need to get out are quite large. I'm not sure how I could clean this up, make it cleaner.